### PR TITLE
Fix regression after pr-10106

### LINKED ISSUE #10159: Investigate/debug how ifBreak with group id works inside group with that id
https://github.com/prettier/prettier/pull/10160#issuecomment-769469692

### PR DESCRIPTION
**Pull request by thorn0**
## Description

Fixes #10159

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/array.js [hunk 1]
```diff
@@ -106,6 +98,16 @@ function printArray(path, options, print) {
...
 
+    const trailingComma = !canHaveTrailingComma
+      ? ""
+      : needsForcedTrailingComma
+      ? ","
+      : !shouldPrintComma(options)
+      ? ""
+      : ifBreak(",", "", {
+          groupId: shouldUseConciseFormatting ? groupId : undefined,
+        });
```
Comment (fisker) [inline src/language-js/print/array.js:109]:
```suggestion
let trailingComma = "";
if (canHaveTrailingComma) {
  if (needsForcedTrailingComma) {
    trailingComma = ",";
  } else if (shouldPrintComma(options)) {
    trailingComma = ifBreak(",", "", {
      groupId: shouldUseConciseFormatting ? groupId : undefined,
    });
  }
}
```

**Comment by thorn0** [comment]
I fixed the issue for non-concise formatting, but it still exists for number arrays:

**Prettier pr-10160**
[Playground link](https://deploy-preview-10160--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEwA6UAEmBmEBOmAFJFAM4yZn5gAKAhjABaYQ6YDaAjAEwDMAGky8+AfRHj+QiTP7jpYiXyVdRAFj48AugEpM6LJgC+GIyAEgIABxgBLaGWSh6+fBADuDfAkcp6AG3d6AE9HCwAjfHowAGs4GABlK2jbKABzZBh8AFc4CzgAW3C4ABMS0oAZenTs+jS4ADECAsY7dOQQemyYCHMQJhgC-wB1Jlt4MmSwOASfcdsAN3HgjrAyMJBUsjh8GFootJbkHADtiwArMgAPACEo2PiE+gK4CtS4Y9O8kEurhNS0v44ABFbIQeCffxnEDJfDbfAdcL0Yr+PpWfCpGDDWwlZjIAAcAAYLOiINthlErB10XB4QsPhYAI5g+D7ay+TpkAC0UDgpVKfW8zNs3n2dSOSBOUO+2wKtkyORlAKBoPBH0lXwsMGR2NxTGQPC1UVs-gBAGEIAUJSBaQBWPrZbYAFWRvil0IWuQAklByrAEmAMTYAIK+hIwYJAyHbIxGIA)
```sh
--parser babel
```

**Input:**
```jsx
{
  for (const srcPath of [123, 123_123_123, 123_123_123_1, 13_123_3123_31_432]) {
  }
}
```

**Output:**
```jsx
{
  for (const srcPath of [
    123, 123_123_123, 123_123_123_1, 13_123_3123_31_432,
  ]) {
  }
}

```

File: src/language-js/print/array.js (hunk 1 above)
Comment (thorn0) [reply src/language-js/print/array.js:109]: My version looks better to me :) Maybe it's better to extract `shouldUseConciseFormatting` to the outer conditional chain though.

Comment (thorn0) [reply src/language-js/print/array.js]: It reads like a chain of simple single-line yes/no questions. I see nothing wrong with it.

Comment (fisker) [reply src/language-js/print/array.js]: Honestly, I never used a ternary nested more than 2 levels, so I never care about https://github.com/prettier/prettier/pull/9559

I can read it, but a little slow than `if`.

**Comment by thorn0** [comment]
I have to admit I don't know how to fix https://github.com/prettier/prettier/pull/10160/files#issuecomment-768334756
Let's merge this as is -so that non-number arrays won't get reformatted - and keep the issue open, shall we?

File: src/language-js/print/array.js (hunk 1 above)
Comment (fisker) [reply src/language-js/print/array.js]: Another thing I don't like in your version, there are two negated conditions, I usually avoid it.

Comment (thorn0) [reply src/language-js/print/array.js]: When I see nested `if`s like that, it gives me an impression that something tricky and intimidating (=== "slower to read") is going on there that can't be expressed as a simple yes/no chain. That said, I prefer nested `if`s when conditions are more complex than in this case. I really appreciate that you're a fan of ESLint (thanks for all those internal rules), but IMHO you're too strict on innocent code structures that are perfectly okay.

**Comment by thorn0** [comment]
Turns out the thing I can't fix (https://github.com/prettier/prettier/pull/10160/files#issuecomment-768334756) happens only in `for-of`, which isn't a common case for number arrays.

**Prettier pr-10160**
[Playground link](https://deploy-preview-10160--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDABAXnQbQKwD6AjAEwDMANOqWceXZdfTQ8VUbS2V0QQCxkS7TvW68x-QQF0A3AB0owBenQAbODHQAjLLjyF9hGu2anGLC82Gtuo3gKHVBDWQoC+ICiAgAHGAEtoAGdkUFQAJ3CIAHcABQiEEJRUVWjUAE8Qry1w1DAAaw0AZR88-ygAc2QYcIBXOC84AFstOAATNvaAGVRK2tQKuAAxCHCm1BgAyuQ0WpgITxAACxgm1QB1Jf94INKwOCLE7f8AN230mbAgrJByoLhwmFjcivHkADMU+68AKyCADwAQrkCsVUE04F1ynAPl8GiA-v8iuUKuoAIq1CDwWGqb4gUrhe7hGZaVCtVSLHzhcowdb+NowJbIAAcAAYvFSIPd1rkfDMqXAiScYV4AI6Y+DPXxJNBBAC0UDg7Xai3CcHF-jVzwGbyQn1x8PuTX81TqhpR6IlML1cK8MDJdIZTKQQhANVQ-lUKIAwhAmrqQIK8ItavcACpkpL6vEneoASSgnVgRTA1L8AEFE0UYOl1Dj7m43EA)
```sh
--parser babel
```

**Input:**
```jsx
a = [5_123, 123_123_123, 123_123_123_1, 13_123_3123_31_432, 13_123_311_311_432];
{
  let b = [55_555_123, 123_123_123, 123_123_123_1, 13_123_3123_31_432, 132_123];
}
```

**Output:**
```jsx
a = [5_123, 123_123_123, 123_123_123_1, 13_123_3123_31_432, 13_123_311_311_432];
{
  let b = [55_555_123, 123_123_123, 123_123_123_1, 13_123_3123_31_432, 132_123];
}

```

**Comment by thorn0** [comment]
Isolated a doc to reproduce the issue. If I replace `group(["const srcPath"])` with `["const srcPath"]` the behavior becomes as expected. Looks like using `ifBreak` with a group id inside a group marked with that id, is an undefined behavior.

```js
[
  "for (",
  group(["const srcPath"]),
  " of ",
  group([
    "[",
    indent([
      softline,
      fill([
        [["123_456_789_123_451_123_123_456"], ","],
        line,
        [["113_123_3123_131_432"], ifBreak(",", "", { groupId: 1 })]
      ]),
    ]),
    softline,
    "]",
  ], {id: 1}),
  ") {",
  hardline,
  "}",
]
```

**Comment by thorn0** [comment]
I hope my comments above don't create a scary impression that this PR is unsafe to merge.

Here what's left unsolved here:

"If a `for-of` loop has a number-only array as iterable, then this line breaks one char earlier than it should."

Not a big deal at all as you can see. Please review.

**Review by fisker — APPROVED** [review]

**Pull request merged**